pkg/kubelet/images/image_manager.go

```diff
@@ -136,12 +185,32 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 	}
 
 	present := imageRef != ""
-	if !shouldPullImage(container, present) {
+
+	var pulledBySecret, ensuredBySecret bool
+	if present {
+		pulledBySecret, ensuredBySecret, err := m.isEnsuredBySecret(imageRef, spec, pullSecrets)
+		if err != nil {
+			klog.V(4).ErrorS(err, "Error get ensured check by secret", "image", image, "imageRef", imageRef, "pulledBySecret", pulledBySecret, "ensuredBySecret", ensuredBySecret)
+		} else {
+			klog.V(5).InfoS("Get ensured check by secret", "image", image, "imageRef", imageRef, "pulledBySecret", pulledBySecret, "ensuredBySecret", ensuredBySecret)
+		}
+	}
+
+	if !shouldPullImage(container, present, pulledBySecret, ensuredBySecret) {
+		// should not pull when pull never, or if present and correctly authenticated
 		if present {
+			if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) && (pulledBySecret && !ensuredBySecret) {
+				// TODO: add integration test for this thrown error message
+				msg := fmt.Sprintf("Container image %q is present with pull policy of %q but does not have the proper auth (image secret) that was used to pull the image", container.Image, container.ImagePullPolicy)
+				m.logIt(ref, v1.EventTypeWarning, events.ErrImageNotEnsured, logPrefix, msg, klog.Warning)
+				return "", msg, ErrImageNotEnsured
```

Comment:
The function returns (imageRef, error message, error), you already have the `ref`, why not to use it while returning instead of ""  ?

Reply:
Yeah, we can return the `imageRef` here and below.